File: src/lab_metrics/labels.py
"""نام metric و labelهای کم‌کاردینالیتی را encode/decode می‌کند."""

from urllib.parse import quote, unquote


def metric_field(name: str, labels: dict[str, object] | None = None) -> str:
    clean_name = _clean(name, 80)
    if not labels:
        return clean_name
    encoded = "&".join(
        f"{quote(_clean(key, 32), safe='')}={quote(dimension(value), safe='')}"
        for key, value in sorted(labels.items())
        if value is not None
    )
    return f"{clean_name}|{encoded}" if encoded else clean_name


def parse_field(value: str | bytes) -> tuple[str, dict[str, str]]:
    text = value.decode() if isinstance(value, bytes) else str(value)
    name, separator, encoded = text.partition("|")
    labels: dict[str, str] = {}
    if separator:
        for pair in encoded.split("&"):
            key, found, item = pair.partition("=")
            if found:
                labels[unquote(key)] = unquote(item)
    return name, labels


def dimension(value: object, limit: int = 80) -> str:
    return _clean(str(value or "unknown"), limit)


def _clean(value: str, limit: int) -> str:
    cleaned = "".join(
        character if character.isalnum() or character in "._-:/" else "_"
        for character in value.strip()
    )
    return (cleaned or "unknown")[:limit]

File: internal/metrics/labels.go
package metrics

import (
	"fmt"
	"net/url"
	"sort"
	"strings"
	"unicode"
)

func Field(name string, labels map[string]any) string {
	cleanName := clean(name, 80)
	if len(labels) == 0 {
		return cleanName
	}
	keys := make([]string, 0, len(labels))
	for key, value := range labels {
		if value != nil {
			keys = append(keys, key)
		}
	}
	sort.Strings(keys)
	pairs := make([]string, 0, len(keys))
	for _, key := range keys {
		pairs = append(pairs, url.QueryEscape(clean(key, 32))+"="+url.QueryEscape(dimension(labels[key])))
	}
	if len(pairs) == 0 {
		return cleanName
	}
	return cleanName + "|" + strings.Join(pairs, "&")
}

func ParseField(value string) (string, map[string]string) {
	name, encoded, found := strings.Cut(value, "|")
	labels := map[string]string{}
	if !found {
		return name, labels
	}
	for _, pair := range strings.Split(encoded, "&") {
		key, item, ok := strings.Cut(pair, "=")
		if !ok {
			continue
		}
		decodedKey, keyErr := url.QueryUnescape(key)
		decodedItem, itemErr := url.QueryUnescape(item)
		if keyErr == nil && itemErr == nil {
			labels[decodedKey] = decodedItem
		}
	}
	return name, labels
}

func dimension(value any) string { return clean(toString(value), 80) }
func toString(value any) string {
	if value == nil {
		return "unknown"
	}
	return fmt.Sprint(value)
}
func clean(value string, limit int) string {
	var builder strings.Builder
	for _, character := range strings.TrimSpace(value) {
		if unicode.IsLetter(character) || unicode.IsDigit(character) || strings.ContainsRune("._-:/", character) {
			builder.WriteRune(character)
		} else {
			builder.WriteByte('_')
		}
	}
	result := builder.String()
	if result == "" {
		result = "unknown"
	}
	runes := []rune(result)
	if len(runes) > limit {
		return string(runes[:limit])
	}
	return result
}
